File: solutions/python/easy/p0283_move_zeroes.py
"""LeetCode 283: Move Zeroes.

Problem: https://leetcode.com/problems/move-zeroes/
"""

from typing import List


class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """Move non-zero values forward, then fill the remaining slots.

        Time: O(n)
        Space: O(1)
        """
        write_index = 0
        for number in nums:
            if number != 0:
                nums[write_index] = number
                write_index += 1
        for index in range(write_index, len(nums)):
            nums[index] = 0

File: solutions/java/easy/P0283_MoveZeroes.java
/**
 * LeetCode 283: Move Zeroes.
 *
 * Problem: https://leetcode.com/problems/move-zeroes/
 */
public class P0283_MoveZeroes {
    public void moveZeroes(int[] nums) {
        int writeIndex = 0;

        for (int number : nums) {
            if (number != 0) {
                nums[writeIndex] = number;
                writeIndex++;
            }
        }

        for (int i = writeIndex; i < nums.length; i++) {
            nums[i] = 0;
        }
    }

    public static void main(String[] args) {
        P0283_MoveZeroes solution = new P0283_MoveZeroes();

        int[] first = new int[] {0, 1, 0, 3, 12};
        solution.moveZeroes(first);
        printArray(first); // Expected: [1, 3, 12, 0, 0]

        int[] second = new int[] {0};
        solution.moveZeroes(second);
        printArray(second); // Expected: [0]
    }

    private static void printArray(int[] nums) {
        System.out.print("[");
        for (int i = 0; i < nums.length; i++) {
            if (i > 0) {
                System.out.print(", ");
            }
            System.out.print(nums[i]);
        }
        System.out.println("]");
    }
}
